base_classes/NXbeam.nxdl.xml

```diff
@@ -43,84 +56,167 @@
     <field name="incident_energy" type="NX_FLOAT" units="NX_ENERGY">
         <doc>Energy on entering beamline component</doc>
         <dimensions rank="1">
-            <dim index="1" value="i"/>
+            <dim index="1" value="m"/>
         </dimensions>
     </field>
     <field name="final_energy" type="NX_FLOAT" units="NX_ENERGY">
         <doc>Energy on leaving beamline component</doc>
         <dimensions rank="1">
-            <dim index="1" value="i"/>
+            <dim index="1" value="m"/>
         </dimensions>
     </field>
     <field name="energy_transfer" type="NX_FLOAT" units="NX_ENERGY">
         <doc>Energy change caused by beamline component</doc>
         <dimensions rank="1">
-            <dim index="1" value="i"/>
+            <dim index="1" value="m"/>
         </dimensions>
     </field>
     <field name="incident_wavelength" type="NX_FLOAT" units="NX_WAVELENGTH">
-        <doc>Wavelength on entering beamline component</doc>
-        <dimensions rank="1">
-            <dim index="1" value="i"/>
-        </dimensions>
+        <doc>
+            In the case of a monchromatic beam this is the scalar
+            wavelength.
+
+            Several other use cases are permitted, depending on the
+            presence or absence of other incident_wavelength_X
+            fields.
+
+            In the case of a polychromatic beam this is an array of
+            length **m** of wavelengths, with the relative weights
+            in ``incident_wavelength_weights``.
+
+            In the case of a monochromatic beam that varies shot-
+            to-shot, this is an array of wavelengths, one for each
+            recorded shot. Here, ``incident_wavelength_weights`` and
+            incident_wavelength_spread are not set.
+
+            In the case of a polychromatic beam that varies shot-to-
+            shot, this is an array of length **m** with the relative
+            weights in ``incident_wavelength_weights`` as a 2D array.
+
+            In the case of a polychromatic beam that varies shot-to-
+            shot and where the channels also vary, this is a 2D array
+            of dimensions **nP** by **m** (slow to fast) with the
+            relative weights in ``incident_wavelength_weights`` as a 2D
+            array.
+
+            Note, :ref:`variants &lt;Design-Variants&gt;` are a good way
+            to represent several of these use cases in a single dataset,
+            e.g. if a calibrated, single-value wavelength value is
+            available along with the original spectrum from which it
+            was calibrated.
+            Wavelength on entering beamline component
```

Comment:
Move sections of this text to top and generally drop the dimnensionality of the other fields, where it's up to the application definition to specify the use of nP, m, both, or neither.

Reply:
After review, it's better to leave this text here.  See comment below.